internal/nearcache/record_store.go

```diff
@@ -116,6 +163,79 @@ func (rs *RecordStore) Invalidate(key interface{}) {
 	rs.incrementInvalidationRequests()
 }
 
+func (rs *RecordStore) doEviction() bool {
+	// port of: com.hazelcast.internal.nearcache.impl.store.AbstractNearCacheRecordStore#doEviction
+	// note that the reference implementation never has withoutMaxSizeCheck == true
+	// checkAvailable doesn't apply
+	if rs.evictionDisabled {
+		return false
+	}
+	return rs.evict()
+}
+
+func (rs *RecordStore) DoExpiration() {
+	// port of: com.hazelcast.internal.nearcache.impl.store.BaseHeapNearCacheRecordStore#doExpiration
+	now := time.Now().UnixMilli()
+	rs.recordsMu.Lock()
+	for k, v := range rs.records {
+		if rs.recordExpired(v, now) {
+			rs.invalidate(k)
+			rs.onExpire()
+		}
+	}
+	rs.recordsMu.Unlock()
+}
+
+func (rs *RecordStore) onEvict(key interface{}, rec *Record, expired bool) {
+	// port of: com.hazelcast.internal.nearcache.impl.store.BaseHeapNearCacheRecordStore#onEvict
+	if expired {
+		rs.incrementExpirations()
+	} else {
+		rs.incrementEvictions()
+	}
+	rs.decrementOwnedEntryCount()
+	rs.decrementOwnedEntryMemoryCost(rs.getTotalStorageMemoryCost(key, rec))
+}
+
+func (rs *RecordStore) tryEvict(candidate evictionCandidate) bool {
+	// port of: com.hazelcast.internal.nearcache.impl.store.HeapNearCacheRecordMap#tryEvict
+	if exists := rs.remove(candidate.Key()); !exists {
+		return false
+	}
+	rs.onEvict(candidate.key, candidate.evictable, false)
+	return true
+}
+
+func (rs *RecordStore) remove(key interface{}) bool {
+	// the key is already in the "made" form, so don't run on rs.makeMapKey on it
+	// assumes rs.recordsMu is locked elsewhere
+	if _, exists := rs.records[key]; !exists {
+		return false
+	}
+	delete(rs.records, key)
+	return true
+}
+
+func (rs *RecordStore) sample(count int) []evictionCandidate {
+	// port of: com.hazelcast.internal.nearcache.impl.store.HeapNearCacheRecordMap#sample
```

Comment:
Is this from https://github.com/hazelcast/hazelcast/blob/14ef0dbefa4a6fa00a50a8873aeea02ec6c3f5e4/hazelcast/src/main/java/com/hazelcast/internal/util/SampleableConcurrentHashMap.java#L300

because from where you link I reached this point. Though, I did not see a similarity between your code and the code I give a link to

Reply:
Adde an explanation at: https://github.com/hazelcast/hazelcast-go-client/pull/807/commits/0d49f98e13b8fc0d4843db1366d353fb6a96a174